tfx/examples/imdb/imdb_utils_native_keras.py

```diff
@@ -0,0 +1,206 @@
+# and change the batch configuration accordingly.
+_DROPOUT_RATE = 0.2
+_EVAL_BATCH_SIZE = 5 
+_HIDDEN_UNITS = 21 
+_LABEL_KEY = "label"
+_LEARNING_RATE = 1e-4
+_MAX_FEATURES = 8000
+_MAX_LEN = 128
+_TRAIN_BATCH_SIZE = 10 
+_TRAIN_EPOCHS = 10
+
+def _gzip_reader_fn(filenames):
+  """Small utility returning a record reader that can read gzip'ed files."""
+  return tf.data.TFRecordDataset(filenames, compression_type='GZIP')
+
+def _tokenize_review(review):
+  """Tokenize the reviews by spliting the reviews, then constructing a
+  vocabulary. Map the words to their frequency index in the vocabulary."""
+  review_sparse = tf.strings.split(tf.reshape(review, [-1])).to_sparse()
+  # tft.apply_vocabulary doesn't reserve 0 for oov words. In order to comply
+  # with convention and use mask_zero in keras.embedding layer, manually set
+  # default value to -1 and add 1 to every index.
+  review_indices = tft.compute_and_apply_vocabulary(
+      review_sparse,
+      default_value=-1,
+      top_k=_MAX_FEATURES)
+  dense = tf.sparse.reset_shape(review_indices, None)
+  dense = tf.sparse.to_dense(review_indices, default_value=-1)
+  # TFX transform expects the transform result to be FixedLenFeature.
+  padding_config = [[0, 0], [0, _MAX_LEN]]
+  dense = tf.pad(
+      dense,
+      padding_config,
+      'CONSTANT',
+      -1)
+
+  padded = tf.slice(dense, [0, 0], [-1, _MAX_LEN])
+  padded += 1
+  return padded
+
+# TFX Transform will call this function.
+def preprocessing_fn(inputs):
+  """tf.transform's callback function for preprocessing inputs.
+
+  Args:
+    inputs: map from feature keys to raw not-yet-transformed features.
+
+  Returns:
+    Map from string feature key to transformed feature operations.
+  """
+  label = inputs['label']
+  text = inputs['text']
+  return {
+      'label': label,
+      'embedding_input': _tokenize_review(text)}
+
+def _input_fn(file_pattern: List[Text],
+              tf_transform_output: tft.TFTransformOutput,
+              batch_size: int = 200) -> tf.data.Dataset:
+  """Generates features and label for tuning/training.
+
+  Args:
+    file_pattern: List of paths or patterns of input tfrecord files.
+    tf_transform_output: A TFTransformOutput.
+    batch_size: representing the number of consecutive elements of returned
+      dataset to combine in a single batch.
+
+  Returns:
+    A dataset that contains (features, indices) tuple where features is a
+      dictionary of Tensors, and indices is a single Tensor of label indices.
+  """
+  transformed_feature_spec = (
+      tf_transform_output.transformed_feature_spec().copy())
+
+  dataset = tf.data.experimental.make_batched_features_dataset(
+      file_pattern=file_pattern,
+      batch_size=batch_size,
+      features=transformed_feature_spec,
+      reader=_gzip_reader_fn,
+      label_key=_LABEL_KEY)
+
+  return dataset
+
+def _build_keras_model() -> keras.Model:
+  """Creates a DNN Keras model for classifying imdb data.
```

Comment:
Is this comment accurate? Maybe add more descriptive pydoc on the model structure?

Also, if this uses modeling technique from elsewhere, consider adding a link too?

Reply:
Nice catch, yes this is bidrectional lstm

Reply:
fixed